streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java

```diff
@@ -115,8 +161,216 @@ public void shutdown() throws IOException {
+        Thread[] streamThreads = new Thread[numThreads];
+        final int numIterations = 500000;
+
+        // create concurrent producer
+        ProducerRunnable producerRunnable = new ProducerRunnable(STREAM_THREE, inputValues, numIterations);
+        Thread producerThread = new Thread(producerRunnable);
+
+        // create three stream threads
+        for (int i = 0; i < numThreads; i++) {
+            streamRunnables[i] = new StreamRunnable(STREAM_THREE, OUTPUT_TOPIC_THREE, i);
+            streamThreads[i] = new Thread(streamRunnables[i]);
+            streamThreads[i].start();
+        }
+        producerThread.start();
+
+        try {
+            waitUntilAtLeastNumRecordProcessed(OUTPUT_TOPIC_THREE, 1);
+
+            for (int i = 0; i < numThreads; i++) {
+                verifyAllKVKeys(streamRunnables, streamRunnables[i].getStream(), inputValuesKeys,
+                    new HashSet(Arrays.asList("word-count-store-" + STREAM_THREE)));
+                verifyAllWindowedKeys(streamRunnables, streamRunnables[i].getStream(), inputValuesKeys,
+                    new HashSet(Arrays.asList("windowed-word-count-store-" + STREAM_THREE)), 0L, WINDOW_SIZE);
+            }
+
+            // kill N-1 threads
+            for (int i = 1; i < numThreads; i++) {
+                streamRunnables[i].close();
+                streamThreads[i].interrupt();
+                streamThreads[i].join();
+            }
+
+            // query from the remaining thread
+            verifyAllKVKeys(streamRunnables, streamRunnables[0].getStream(), inputValuesKeys,
+                new HashSet(Arrays.asList("word-count-store-" + STREAM_THREE)));
+            verifyAllWindowedKeys(streamRunnables, streamRunnables[0].getStream(), inputValuesKeys,
+                new HashSet(Arrays.asList("windowed-word-count-store-" + STREAM_THREE)), 0L, WINDOW_SIZE);
+        } finally {
+            for (int i = 0; i < numThreads; i++) {
+                if (!streamRunnables[i].isClosed()) {
+                    streamRunnables[i].close();
+                    streamThreads[i].interrupt();
+                    streamThreads[i].join();
+                }
+            }
+            producerRunnable.shutdown();
+            producerThread.interrupt();
+            producerThread.join();
+        }
+    }
+
+    @Test
+    public void concurrentAccesses() throws Exception {
+
+        final int numIterations = 500000;
+
+        ProducerRunnable producerRunnable = new ProducerRunnable(STREAM_CONCURRENT, inputValues, numIterations);
+        Thread producerThread = new Thread(producerRunnable);
+        kafkaStreams = createCountStream(STREAM_CONCURRENT, OUTPUT_TOPIC_CONCURRENT, streamsConfiguration);
+        kafkaStreams.start();
+        producerThread.start();
+
+        try {
+            waitUntilAtLeastNumRecordProcessed(OUTPUT_TOPIC_CONCURRENT, 1);
+
+            final ReadOnlyKeyValueStore<String, Long>
```

Comment:
nit: rename `myCount` to `keyValueStore` to be consistent with the `windowStore`